### ISSUE TITLE
Add rule “too many positional arguments in function definition”

### ISSUE THREAD

**Issue by flying-sheep** [astral-sh/ruff#8946]
Since https://github.com/astral-sh/ruff/pull/4329, PLR0913 is useless to us. We need a rule that doesn’t restrict the number of total arguments, but the number of non-keyword-only arguments. See also https://github.com/pylint-dev/pylint/issues/9099

I’m happy to do a PR, but I don’t know how I should go about the rule identifier: Should I go for a new RUFF rule, or is there a way I can reserve a new code like e.g. PLR0914 in Pylint for this even if I don’t plan to make a PR for Pylint?

**Comment by flying-sheep** [comment]
It’s pretty simple: Good APIs don’t have many positional parameters.

Most functions take 1–2 “main objects” to operate on and a few parameters that modify how they operate, e.g. `zip(a: Iterable, b: Iterable, strict: bool)`. Then these main objects should be able to be specified as positional parameters and the rest as keyword-only.

There are few exceptions where 4 or more positional parameters make sense, e.g. `rgba(1.0, 0.5, 0.3, 1.0)` is a very well known conventional order of 4 floats/bytes.

**Comment by flying-sheep** [comment]
positional vs total arguments. you can configure:

- maximum number of positional arguments: 3
- maximum number of arguments: 6

Then this would be legal:
```py
def (a, b, c, *, x, y, z): ...
```

and this wouldn‘t (because while it has <=6 arguments, it has >3 positional ones)
```py
def (a, b, c, x, y, z): ...
```

**Comment by flying-sheep** [comment]
Exactly: https://docs.python.org/3/tutorial/controlflow.html#special-parameters

**Comment by flying-sheep** [comment]
Always happy to do so!

**Comment by Pierre-Sassoulas** [comment]
@buhtz did you read pylint's 3.3.1's doc or the latest ? And what helped you understand the concept so we can put it in the doc ? Sorry for not removing the link to this issue, it was relevant at a time when the pylint check was not implemented yet, but not anymore.
